File: sdk/java/src/main/java/com/ouc/tcp/sdk/rdt1/Rdt1Receiver.java
package com.ouc.tcp.sdk.rdt1;

import com.ouc.tcp.sdk.Packet;
import com.ouc.tcp.sdk.SystemContext;
import com.ouc.tcp.sdk.TransportProtocol;

public final class Rdt1Receiver implements TransportProtocol {
    @Override
    public void init(SystemContext ctx) {
        ctx.log("Java RDT1 receiver ready");
    }

    @Override
    public void onPacket(SystemContext ctx, Packet packet) {
        ctx.log("RDT1 receiver delivering " + packet.getPayload().length + " bytes");
        ctx.deliverData(packet.getPayload());
    }

    @Override
    public void onTimer(SystemContext ctx, int timerId) {
        ctx.log("RDT1 receiver ignores timer " + timerId);
    }

    @Override
    public void onAppData(SystemContext ctx, byte[] data) {
        ctx.log("RDT1 receiver ignores outbound data of len " + data.length);
    }
}

File: sdk/cpp/src/rdt1_receiver.cpp
#include "tcp_lab/sdk.hpp"

using namespace tcp_lab::sdk;

class Rdt1Receiver final : public Protocol {
  public:
    void init() override {
        log("C++ RDT1 receiver ready");
    }

    void on_packet(const TcpHeader&, const std::vector<uint8_t>& payload) override {
        log("RDT1 receiver delivering " + std::to_string(payload.size()) + " bytes");
        deliver_data(payload);
    }
};

TCP_LAB_REGISTER_PROTOCOL(Rdt1Receiver)
